File: ios/RNGiphyCache.m
//
//  RNGiphyCache.m
//  GiphyUi
//
//  Created by Jesse Onolememen on 27/03/2020.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>

#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include(“RCTBridgeModule.h”)
#import "RCTBridgeModule.h"
#else
#import "React/RCTBridgeModule.h"
#endif

@interface RCT_EXTERN_MODULE(RNGiphyCache, NSObject)

RCT_EXTERN_METHOD(clear)
RCT_EXTERN_METHOD(setDiskCacheByteLimit: (int)limit)
RCT_EXTERN_METHOD(setSetting: (NSString *)setting)
RCT_EXTERN_METHOD(downalodAssetData: (NSString *)url resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock) reject)

@end

File: ios/RNGiphyCache.swift
//
//  GiphyCache.swift
//  GiphyUi
//
//  Created by Jesse Onolememen on 27/03/2020.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation
import GiphyCoreSDK
import GiphyUISDK

@objc(RNGiphyCache)
class RNGiphyCache: NSObject {
  @objc func setDiskCacheByteLimit(_ limit: Int) {
    GPHCache.shared.diskCacheByteLimit = UInt(limit)
  }
  
  @objc func setCacheSetting(_ setting: String) {
    switch setting {
    case "memoryOnly":
      GPHCache.shared.setting = .memoryOnly
    case "diskOnly":
      GPHCache.shared.setting = .diskOnly
    default:
      break
    }
  }
  
  @objc func clearCache() {
    GPHCache.shared.clear()
  }
  
  @objc func downalodAssetData(_ url: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    GPHCache.shared.downloadAsset(url) { (image, error) in
      if let image = image, let data = image.pngData() {
        resolve([
          "data": data.base64EncodedString(),
          "width": image.size.width,
          "height": image.size.height,
        ])
      } else {
        reject("GiphyCacheDownloadAssetError", error?.localizedDescription, error)
      }
    }
  }
}
